lib/IDE/IDETypeChecking.cpp

```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
   return Scratch;
 }
 
+/// This walker will traverse the AST and report types for every variable declaration.
+class VariableTypeCollector: public SourceEntityWalker {
+private:
+  SourceManager &SM;
+  unsigned int BufferId;
+  std::vector<VariableTypeInfo> &Results;
+
+  /// This is where we print all types.
+  llvm::raw_ostream &OS;
+
+  /// Map from a printed type to the offset in OS where the type starts.
+  llvm::StringMap<uint32_t> TypeOffsets;
+
+  /// Specified by the client whether we should canonicalize types before printing.
+  const bool CanonicalType;
+
+  /// Find an exicisting offset in the type buffer, otherwise print the type.
+  std::pair<uint32_t, uint32_t> getTypeOffsets(StringRef PrintedType) {
+    auto It = TypeOffsets.find(PrintedType);
+    if (It == TypeOffsets.end()) {
+      TypeOffsets[PrintedType] = OS.tell();
+      OS << PrintedType << '\0';
+    }
+    return {TypeOffsets[PrintedType], PrintedType.size()};
+  }
+public:
+  VariableTypeCollector(SourceFile &SF,
+                        std::vector<VariableTypeInfo> &Results,
+                        bool CanonicalType,
```

Comment:
The `ExpressionTypeInfo` had one, so I added it for consistency. Perhaps in the future it might be nice to have an option for displaying inlay hints with fully qualified types?

Reply:
Honestly, I don’t think there’s much value in using fully-qualified types for inlay hints, they are just too verbose. And we can always add it later if we should discover that it’s needed.